clients/grailfiltersegments/filtersegments.go

```diff
@@ -0,0 +1,228 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grailfiltersegments
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/clients/grailfiltersegements"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/rest"
+	"github.com/go-logr/logr"
+)
+
+const bodyReadErrMsg = "unable to read API response body"
+
+type Response = api.Response
+
+func NewClient(client *rest.Client) *Client {
+	c := &Client{
+		client: grailfiltersegements.NewClient(client),
+	}
+	return c
+}
+
+// Client can be used to interact with the Automation API
+type Client struct {
+	client client
+}
+
+//go:generate mockgen -source filtersegments.go -package=grailfiltersegments -destination=client_mock.go
+type client interface {
+	Get(ctx context.Context, id string, ro rest.RequestOptions) (*http.Response, error)
+	List(ctx context.Context) (*http.Response, error)
+	Create(ctx context.Context, data []byte) (*http.Response, error)
+	Update(ctx context.Context, id string, data []byte, ro rest.RequestOptions) (*http.Response, error)
+	Delete(ctx context.Context, id string) (*http.Response, error)
+}
+
+var _ client = (*grailfiltersegements.Client)(nil)
+
+func (c Client) Get(ctx context.Context, id string) (Response, error) {
+	if id == "" {
+		return Response{}, errors.New("missing required id")
+	}
+	resp, err := c.client.Get(ctx, id, rest.RequestOptions{CustomShouldRetryFunc: rest.RetryIfTooManyRequests})
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to get filtersegment resource with id %s: %w", id, err)
+	}
+	if resp != nil && resp.Body != nil {
+		defer resp.Body.Close()
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		logr.FromContextOrDiscard(ctx).Error(err, bodyReadErrMsg)
+		return Response{}, api.NewAPIErrorFromResponseAndBody(resp, body)
+	}
+
+	if !rest.IsSuccess(resp) {
+		return Response{}, api.NewAPIErrorFromResponseAndBody(resp, body)
+	}
+
+	return api.NewResponseFromHTTPResponseAndBody(resp, body), nil
+}
+
+func (c Client) List(ctx context.Context) (Response, error) {
+	resp, err := c.client.List(ctx)
+	if err != nil {
+		return Response{}, fmt.Errorf("failed to list filtersegments resources: %w", err)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		logr.FromContextOrDiscard(ctx).Error(err, bodyReadErrMsg)
+		return Response{}, api.NewAPIErrorFromResponseAndBody(resp, body)
+	}
+	if !rest.IsSuccess(resp) {
+		return Response{}, api.NewAPIErrorFromResponseAndBody(resp, body)
+	}
+	{
+		var tmp map[string]any
+		if err = json.Unmarshal(body, &tmp); err != nil {
+			return Response{}, api.NewAPIErrorFromResponseAndBody(resp, body)
+		}
+		if body, err = json.Marshal(tmp["filterSegments"]); err != nil {
```

Comment:
Could you please eloberate what we're doing here exactly? Are we removing everything except the filterSegments? If yes, why? What else is in there?

Reply:
Moved to separate function. It is also done in description of the function